File: src/api/exercises.ts
import { invoke } from "@tauri-apps/api/core";
import type { Exercise, ExerciseMeta, ExerciseMuscleInput, ExerciseReferences, ExerciseSearchFilters, ExerciseSearchResult } from "../types/exercise";

export const exercisesApi = {
  list: () => invoke<Exercise[]>("list_exercises"),

  get: (id: string) =>
    invoke<Exercise>("get_exercise", { id }),

  search: (filters: ExerciseSearchFilters) =>
    invoke<ExerciseSearchResult>("search_exercises", { filters }),

  create: (
    name: string,
    notes: string | null,
    tags: string[] = [],
    meta?: ExerciseMeta,
    muscles?: ExerciseMuscleInput[],
    poseTypes?: string[],
  ) =>
    invoke<Exercise>("create_exercise", {
      name,
      notes,
      tags,
      meta: meta ?? null,
      muscles: muscles ?? null,
      poseTypes: poseTypes ?? null,
    }),

  update: (
    id: string,
    name: string,
    notes: string | null,
    tags: string[] = [],
    meta?: ExerciseMeta,
    muscles?: ExerciseMuscleInput[],
    poseTypes?: string[],
  ) =>
    invoke<Exercise>("update_exercise", {
      id,
      name,
      notes,
      tags,
      muscles: muscles ?? null,
      meta: meta ?? null,
      poseTypes: poseTypes ?? null,
    }),

  getReferences: (id: string) =>
    invoke<ExerciseReferences>("get_exercise_references", { id }),

  delete: (id: string, confirmed: boolean) =>
    invoke<void>("delete_exercise", { id, confirmed }),
};

File: src-tauri/src/commands/exercises.rs
use sqlx::SqlitePool;
use tauri::State;
use crate::{
    domain::{exercise, types::{Exercise, ExerciseMeta, ExerciseMuscleInput, ExerciseReferences}},
    error::AppError,
};

#[tauri::command]
pub async fn list_exercises(pool: State<'_, SqlitePool>) -> Result<Vec<Exercise>, AppError> {
    exercise::list(&pool).await
}

#[tauri::command]
pub async fn create_exercise(
    pool: State<'_, SqlitePool>,
    name: String,
    notes: Option<String>,
    tags: Vec<String>,
    meta: Option<ExerciseMeta>,
    muscles: Option<Vec<ExerciseMuscleInput>>,
) -> Result<Exercise, AppError> {
    exercise::create(
        &pool,
        &name,
        notes.as_deref(),
        &tags,
        meta.as_ref(),
        muscles.as_deref(),
    )
    .await
}

#[tauri::command]
pub async fn update_exercise(
    pool: State<'_, SqlitePool>,
    id: String,
    name: String,
    notes: Option<String>,
    tags: Vec<String>,
    muscles: Option<Vec<ExerciseMuscleInput>>,
    meta: Option<ExerciseMeta>,
) -> Result<Exercise, AppError> {
    exercise::update(
        &pool,
        &id,
        &name,
        notes.as_deref(),
        &tags,
        meta.as_ref(),
        muscles.as_deref(),
    )
    .await
}

#[tauri::command]
pub async fn get_exercise_references(
    pool: State<'_, SqlitePool>,
    id: String,
) -> Result<ExerciseReferences, AppError> {
    exercise::get_references(&pool, &id).await
}

#[tauri::command]
pub async fn delete_exercise(
    pool: State<'_, SqlitePool>,
    id: String,
    confirmed: bool,
) -> Result<(), AppError> {
    exercise::delete_with_unlink(&pool, &id, confirmed).await
}
